File: rust-sdk/tests/common/planners/sponsored_buy.rs
//! Sponsored swap settlement — mirrors [`tests/sponsored_buy.ts`](../../../../tests/sponsored_buy.ts).

use ifx_sdk::expr;
use ifx_sdk::patched_cpi::{
    build_structured_cpi, frame_value, structured_system_transfer, system_transfer_template,
};
use ifx_sdk::scratch::FrameScratch;
use ifx_sdk::ScratchError;
use solana_sdk::instruction::Instruction;
use solana_sdk::pubkey::Pubkey;
use solana_system_interface::instruction as system_instruction;

#[derive(Clone, Debug)]
pub struct SponsoredBuyAccounts {
    pub sponsor: Pubkey,
    pub user: Pubkey,
    pub pool: Pubkey,
    pub user_ata: Pubkey,
}

#[derive(Clone, Debug)]
pub struct SponsoredBuyParams {
    pub tx_sig_fee: u64,
}

pub fn plan_sponsored_buy_instructions(
    scratch: &mut FrameScratch,
    accts: &SponsoredBuyAccounts,
    params: &SponsoredBuyParams,
    mock_swap_lamports: u64,
    idempotent_ata_create: Instruction,
) -> Result<Vec<Instruction>, ScratchError> {
    let SponsoredBuyAccounts {
        sponsor,
        user,
        pool,
        user_ata,
    } = *accts;
    let tx_sig_fee = params.tx_sig_fee;

    let mut out = Vec::new();
    out.push(scratch.ix_reset());

    let mut let_baseline = scratch.let_builder();
    let user_lamports_baseline = let_baseline.lamports(user)?;
    let ata_lamports_baseline = let_baseline.lamports(user_ata)?;
    out.push(let_baseline.build_ix()?);

    out.push(idempotent_ata_create);

    let mut let_ata = scratch.let_builder();
    let ata_lamports_after_create = let_ata.lamports(user_ata)?;
    let ata_cost = let_ata.let_eval(expr::sub(
        expr::r(&ata_lamports_after_create),
        expr::r(&ata_lamports_baseline),
    ))?;
    out.push(let_ata.build_ix()?);

    out.push(system_instruction::transfer(
        &pool,
        &user,
        mock_swap_lamports,
    ));

    let mut let_post_swap = scratch.let_builder();
    let user_lamports_after_swap = let_post_swap.lamports(user)?;
    let settle = let_post_swap.let_eval(expr::add(
        expr::r(&ata_cost),
        expr::u64(tx_sig_fee),
    ))?;
    let buy_lamports = let_post_swap.let_eval(expr::sub(
        expr::sub(
            expr::r(&user_lamports_after_swap),
            expr::r(&user_lamports_baseline),
        ),
        expr::r(&settle),
    ))?;
    out.push(let_post_swap.build_ix()?);

    out.push(scratch.ix_assert(&expr::ge(
        expr::sub(
            expr::r(&user_lamports_after_swap),
            expr::r(&user_lamports_baseline),
        ),
        expr::add(expr::r(&ata_cost), expr::u64(tx_sig_fee)),
    ))?);

    let settle_xfer = build_structured_cpi(
        &system_transfer_template(user, sponsor),
        structured_system_transfer(frame_value(&settle)),
    )?;
    out.push(scratch.ix_cpi(&settle_xfer)?);

    let buy_xfer = build_structured_cpi(
        &system_transfer_template(user, pool),
        structured_system_transfer(frame_value(&buy_lamports)),
    )?;
    out.push(scratch.ix_cpi(&buy_xfer)?);

    Ok(out)
}

File: go-sdk/examples/sponsored_buy.go
package examples

import (
	"github.com/gagliardetto/solana-go"
	"github.com/gagliardetto/solana-go/programs/system"
	"github.com/ifx-run/ifx/go-sdk/expr"
	"github.com/ifx-run/ifx/go-sdk/scratch"
	"github.com/ifx-run/ifx/go-sdk/structuredcpi"
)

// SponsoredBuyAccounts wire accounts for sponsored swap settlement.
type SponsoredBuyAccounts struct {
	Sponsor solana.PublicKey
	User    solana.PublicKey
	Pool    solana.PublicKey
	UserATA solana.PublicKey
}

// SponsoredBuyParams planner parameters.
type SponsoredBuyParams struct {
	TxSigFee uint64
}

// PlanSponsoredBuyInstructions mirrors tests/sponsored_buy.ts and rust-sdk planners/sponsored_buy.rs.
func PlanSponsoredBuyInstructions(
	s *scratch.FrameScratch,
	accts SponsoredBuyAccounts,
	params SponsoredBuyParams,
	mockSwapLamports uint64,
	idempotentATACreate solana.Instruction,
) ([]solana.Instruction, error) {
	txSigFee := params.TxSigFee
	out := []solana.Instruction{s.IxReset()}

	letBaseline := s.LetBuilder()
	userLamportsBaseline, err := letBaseline.Lamports(accts.User)
	if err != nil {
		return nil, err
	}
	ataLamportsBaseline, err := letBaseline.Lamports(accts.UserATA)
	if err != nil {
		return nil, err
	}
	baselineIx, err := letBaseline.BuildIx()
	if err != nil {
		return nil, err
	}
	out = append(out, baselineIx)

	out = append(out, idempotentATACreate)

	letAta := s.LetBuilder()
	ataLamportsAfterCreate, err := letAta.Lamports(accts.UserATA)
	if err != nil {
		return nil, err
	}
	ataCost, err := letAta.LetEval(expr.Sub(
		expr.Ref(ataLamportsAfterCreate.Index),
		expr.Ref(ataLamportsBaseline.Index),
	))
	if err != nil {
		return nil, err
	}
	ataIx, err := letAta.BuildIx()
	if err != nil {
		return nil, err
	}
	out = append(out, ataIx)

	out = append(out, system.NewTransferInstruction(mockSwapLamports, accts.Pool, accts.User).Build())

	letPostSwap := s.LetBuilder()
	userLamportsAfterSwap, err := letPostSwap.Lamports(accts.User)
	if err != nil {
		return nil, err
	}
	settle, err := letPostSwap.LetEval(expr.Add(
		expr.Ref(ataCost.Index),
		expr.U64(txSigFee),
	))
	if err != nil {
		return nil, err
	}
	buyLamports, err := letPostSwap.LetEval(expr.Sub(
		expr.Sub(
			expr.Ref(userLamportsAfterSwap.Index),
			expr.Ref(userLamportsBaseline.Index),
		),
		expr.Ref(settle.Index),
	))
	if err != nil {
		return nil, err
	}
	postSwapIx, err := letPostSwap.BuildIx()
	if err != nil {
		return nil, err
	}
	out = append(out, postSwapIx)

	assertIx, err := s.IxAssert(expr.Ge(
		expr.Sub(
			expr.Ref(userLamportsAfterSwap.Index),
			expr.Ref(userLamportsBaseline.Index),
		),
		expr.Add(expr.Ref(ataCost.Index), expr.U64(txSigFee)),
	))
	if err != nil {
		return nil, err
	}
	out = append(out, assertIx)

	settleBuilt, err := structuredcpi.StructuredCpi(
		system.NewTransferInstruction(0, accts.User, accts.Sponsor).Build(),
		structuredcpi.StructuredCpiPatch.SystemTransfer(structuredcpi.AsFrameValue(settle)),
	)
	if err != nil {
		return nil, err
	}
	settleXfer, err := settleBuilt.Build(nil)
	if err != nil {
		return nil, err
	}
	settleCpi, err := s.IxCpi(settleXfer)
	if err != nil {
		return nil, err
	}
	out = append(out, settleCpi)

	buyBuilt, err := structuredcpi.StructuredCpi(
		system.NewTransferInstruction(0, accts.User, accts.Pool).Build(),
		structuredcpi.StructuredCpiPatch.SystemTransfer(structuredcpi.AsFrameValue(buyLamports)),
	)
	if err != nil {
		return nil, err
	}
	buyXfer, err := buyBuilt.Build(nil)
	if err != nil {
		return nil, err
	}
	buyCpi, err := s.IxCpi(buyXfer)
	if err != nil {
		return nil, err
	}
	out = append(out, buyCpi)

	return out, nil
}
